python/taichi/lang/kernel_arguments.py

```diff
@@ -36,6 +36,29 @@ def extract(self, x):
 """
 
 
+class ArgAnyArray:
+    """Type annotation for arbitrary arrays, including external arrays and Taichi ndarrays."""
+    def extract(self, x):
+        return to_taichi_type(x.dtype), len(x.shape)
+
+
+any_arr = ArgAnyArray
+"""Alias for :class:`~taichi.lang.kernel_arguments.ArgAnyArray`.
+
+Example::
+
+    >>> @ti.kernel
+    >>> def to_numpy(x: ti.any_arr(), y: ti.any_arr()):
+    >>>     for i in range(n):
+    >>>         x[i] = y[i]
+    >>>
+    >>> y = ti.ndarray(ti.f64, shape=n)
```

Comment:
Nit: if `ti.ndarray` will be a first class citizen in the future, shall we also name the annotation accordingly? 

I was thinking about `ti.annotation.ndarray()` as an example but that might require a bit refactor. 

`ti.ext_arr()` doesn't look like a type annotation from its name :P

Reply:
Nice catch - I agree that the current annotations (`ti.template()` for fields, `ti.ext_arr()` for external arrays) are not straightforward to users. Instead, they are more like directives for compilers regarding which information is available at compile time/runtime. As for `ti.any_arr()` here, the final name is still under discussion but we are aiming for a general annotation which accepts external arrays, Taichi ndarrays and dense Taichi fields. Maybe we need a better naming system in the future.

Reply:
Yup agreed. I feel like this doesn't have to be resolved in this PR, but we probably need to keep this in mind to make it more straightforward to users before we officially release ndarray. :D

Reply:
Sure :-)